src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1006,251 +1047,375 @@ RouterSelectTask(Query *originalQuery, Query *query)
+
+	Assert(list_length(shardIntervalList) == 1);
+	shardInterval = (ShardInterval *) linitial(shardIntervalList);
+
+	shardId = shardInterval->shardId;
+	relationId = shardInterval->relationId;
+	relationName = get_rel_name(relationId);
+	AppendShardIdToName(&relationName, shardId);
+
+	schemaId = get_rel_namespace(relationId);
+	schemaName = get_namespace_name(schemaId);
+
+	/* XXX: why is that a good idea? */
+	if (strncmp(schemaName, "public", NAMEDATALEN) == 0)
 	{
-		return false;
+		schemaName = NULL;
 	}
 
+	ModifyRangeTblExtraData(newRte, CITUS_RTE_SHARD, schemaName,
+							relationName,
+							NIL);
+
+	return false;
+}
+
+
+/*
+ * RouterQueryJob creates a Job for the specified query to execute the
+ * provided single shard select task.
+ */
+static Job *
+RouterQueryJob(Query *query, Task *task, List *placementList)
+{
+	Job *job = NULL;
+	List *taskList = NIL;
+	TaskType taskType = task->taskType;
+
 	/*
-	 * We need to make sure there is at least one shard for this hash partitioned
-	 * query to be router plannable. We can not prepare a router plan if there
-	 * are no shards.
+	 * We send modify task to the first replica, otherwise we choose the target shard
+	 * according to task assignment policy. Placement list for select queries are
+	 * provided as function parameter.
 	 */
-	shardCount = ShardIntervalCount(distributedTableId);
-	if (shardCount == 0)
+	if (taskType == MODIFY_TASK)
 	{
-		return false;
+		taskList = FirstReplicaAssignTaskList(list_make1(task));
 	}
+	else
+	{
+		Assert(placementList != NIL);
 
-	return true;
+		task->taskPlacementList = placementList;
+		taskList = list_make1(task);
+	}
+
+	job = CitusMakeNode(Job);
+	job->dependedJobList = NIL;
+	job->jobId = INVALID_JOB_ID;
+	job->subqueryPushdown = false;
+	job->jobQuery = query;
+	job->taskList = taskList;
+
+	return job;
 }
 
 
 /*
- * ColumnMatchExpressionAtTopLevelConjunction returns true if the query contains an exact
- * match (equal) expression on the provided column. The function returns true only
- * if the match expression has an AND relation with the rest of the expression tree.
+ * MultiRouterPlannableQuery returns true if given query can be router plannable.
+ * The query is router plannable if it is a select query issued on a hash
+ * partitioned distributed table, and it has a exact match comparison on the
+ * partition column. This feature is enabled if task executor is set to real-time
  */
-static bool
-ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
+bool
+MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+						  RelationRestrictionContext *restrictionContext)
 {
-	if (node == NULL)
+	CmdType commandType = query->commandType;
+	ListCell *relationRestrictionContextCell = NULL;
+
+	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
+		commandType == CMD_DELETE)
 	{
-		return false;
+		return true;
 	}
 
-	if (IsA(node, OpExpr))
+	/* FIXME: I tend to think it's time to remove this */
```

Comment:
On 2016-07-19 09:56:14 -0700, Murat Tuncer wrote:
> >  	}
> >  
> > -	if (IsA(node, OpExpr))
> > +	/* FIXME: I tend to think it's time to remove this */
> 
> we won't be able to reach a conclusion here. Let's ask @ozgune to make the call on this.

Just leave the fixme there, and let's finish this PR.